Add variable type sanity checks in FEProblemBase::addXYZ() routines.

jwpeterson opened the pull request:
This fixes the problem pointed out by @YaqiWang in #11021, and adds tests to make sure that we can keep detecting the error.

jwpeterson committed:
Add variable type sanity checks in FEProblemBase::addXYZ() routines.

* addKernel()
* addBoundaryCondition()
* addAuxKernel()
* addScalarKernel()
* addNodalKernel()
* addConstraint()
* addAuxScalarKernel()
* addDiracKernel()
* addDGKernel()
* addInterfaceKernel()

Refs #11021.

jwpeterson committed:
Fix misc/check_error/calling_wrong_feproblem_method test.

This test needed to be modified in a non-essential way so that it
continues to produce the same error message now that the new sanity
checks are in place. Also updated the comment about what we are doing
in the test so that it is a bit more clear.

Refs #11021.

jwpeterson committed:
Add RunException tests to verify new sanity checks are working.

Refs #11021.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/167083/) on 3fc5852 wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/11039/site)

This comment will be updated on new commits.

andrsd commented on framework/src/base/FEProblemBase.C:
```diff
@@ -2081,6 +2142,13 @@ FEProblemBase::addInterfaceKernel(const std::string & interface_kernel_name,
+
+  // Check that "variable" is in the NonlinearSystem.
+  if (!_nl->hasVariable(parameters.get<NonlinearVariableName>("variable")))
+    mooseError("Cannot add InterfaceKernel for variable ",
```
In all `mooseErrors` you want to use the object name - it is coming in the `name` parameter - so that people know which block triggered the error. I suggest something like:
```
mooseError(name, ": error msg here");
```
Especially useful when you have tonnes of objects in your input file.

andrsd commented on test/tests/misc/check_error/tests:
```diff
@@ -343,7 +343,7 @@
     type = 'RunException'
     input = 'missing_req_par_moose_obj_test.i'
-    expect_err = "missing required parameter 'Kernels/diff/variable'"
+    expect_err = "Cannot add Kernel for variable , it is not a nonlinear variable"
```
This does not seem right. The test tests that if people forget a variable parameter, we tell them to add it.  The new error message does not really implies that.   When is the old error message triggered?

jwpeterson replied:
Hmm... "variable" is a `RequiredParam` on `KernelBase`, so if the user forgets to provide it, the original error message is triggered in `InputParameters::checkParams()`, which must be happening after the new sanity check. I suppose I could change the sanity check to something involving `parameters.isParamValid("variable")` as well?

andrsd commented on test/tests/misc/check_error/tests:
```diff
@@ -616,4 +616,55 @@
+  [./aux_kernel_with_var]
+    type = 'RunException'
+    input = 'aux_kernel_with_var.i'
+    expect_err = "Cannot add AuxKernel for variable u, it is not an AuxVariable"
```
probably `it is not an auxiliary variable`, like we spell out `nonlinear variable`.

andrsd commented on test/tests/misc/check_error/tests:
```diff
@@ -616,4 +616,55 @@
+  [./scalar_kernel_with_var]
+    type = 'RunException'
+    input = 'scalar_kernel_with_var.i'
+    expect_err = "Cannot add ScalarKernel for variable u, it is not a Scalar variable"
```
Should `Scalar` be capitalized?

YaqiWang commented:
Do you have similar check for aux kernels where the variable needs to be on aux system?

jwpeterson commented:
> Do you have similar check for aux kernels where the variable needs to be on aux system?

Yes, look at the patch....

jwpeterson committed:
Add 'name' to the error messages.

This way people can see which block of the input file actually
triggered the error. Note that this did not require an update to the
expect_err fields of the various tests since the regex was not
required to start at the beginning of the line...

Refs #11039.

jwpeterson committed:
Fix error message for AuxKernel with nonlinear variable.

jwpeterson committed:
Fix capitalization of SCALAR in error message.

jwpeterson committed:
Change missing required param test slightly.

We are now omitting the 'type' parameter for this Kernel. This way, we are
still testing the same type of error as before (missing parameter) instead
of triggering an earlier sanity check that the variable provided to the
Kernel is actually on the nonlinear system.

Refs #11039.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/167914/) on c2bc9de wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11039/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11039/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 948bf75cc88875c8f252442822cc2fc4460cd505`

jwpeterson committed:
Fix one more capitialization.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/167925/) on eb64bcb wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11039/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11039/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 948bf75cc88875c8f252442822cc2fc4460cd505`

jwpeterson committed:
Apply clang-format patch from CIVET.
